lib/coll.gi

```diff
@@ -246,15 +246,35 @@ InstallMethod( RepresentativeSmallest,
 ##  an enumerator of <C> and selects a random element of this list using the
 ##  function `RandomList', which is a pseudo random number generator.
 ##
-InstallGlobalFunction( RandomList, function(list)
-  return list[Random(GlobalMersenneTwister, 1, Length(list))];
+
+# RandomList is not a Method to avoid the (often expensive) cost of
+# method dispatch for lists
+InstallGlobalFunction( RandomList, function(args...)
+  local len, source, list;
+  len := Length(args);
+  if len = 1 then
+    source := GlobalMersenneTwister;
+    list := args[1];
+  elif len = 2 then
+    source := args[1];
+    list := args[2];
+  else
+    Error(" Usage: RandomList([rs], list))");
+  fi;
+
+  return list[Random(source, 1, Length(list))];
 end );
-InstallMethod( Random, "for a (finite) collection",
-    [ IsCollection and IsFinite ],
+
+# Use -SUM_FLAGS, as this should always be the last choice method.
+# In particular, we want Random(SomeRandomSource, IsList) to come higher
+# for lists, to avoid an infinite loop.
+InstallMethodWithRandomSource( Random, "for a (finite) collection",
+    [ IsCollection and IsFinite ], -SUM_FLAGS,
```

Comment:
I would avoid extra ranking of methods whenever possible. And the `-SUM_FLAGS` does not seem needed here. E.g., in current master branch:

```
gap> DeclareCategory("IsDice", IsComponentObjectRep and IsDomain);
gap> dice := Objectify(NewType(NewFamily(""), IsDice), rec());
<object>
gap> InstallMethod(Enumerator, [IsDice], d-> [1..6]);
gap> 
gap> InstallMethod(Random, ["IsRandomSource", "IsCollection and IsFinite"],
> function(rs, coll) return Random(Enumerator(coll)); end);
gap> Random(dice);
2
```

Reply:
Firstly, there is a small bug here, sorry.

This shows the issue I'm trying to fix. Given the following code, then `Random(GlobalRandomSource,[1..10])` enters an infinite loop, because this newly installed method ranks more highly than the `InstallMethod(Random, [IsGAPRandomSource, IsList],...` in random.gi.

```
InstallMethod( Random, "for a source and a (finite) collection",
    [ IsRandomSource, IsCollection and IsFinite ],
    {rs,C} -> RandomList(rs, Enumerator( C ) ) );


MakeReadWriteGlobal("RandomList");
RandomList := function(args...)
  local len, source, list;
  len := Length(args);
  if len = 1 then
    source := GlobalMersenneTwister;
    list := args[1];
  elif len = 2 then
    source := args[1];
    list := args[2];
  else
    Error(" Usage: RandomList([rs], list))");
  fi;

  return list[Random(source, 1, Length(list))];
end;
```